Add support for generating noise images with an offset

radishes opened the pull request:
**Problem**
`OpenSimplexNoise.get_image()` and `NoiseTexture` can only generate noise from coordinates in the range (0,0) to (image_width, image_height). The result is that these classes can't be used to generate texture data across a larger space. My use case is that I want to generate noise data at arbitrary coordinates so that I can use the data for creating a procedural world, and passing to shaders.

**Workaround**
It is possible to achieve the desired outcome in gdscript. I can generate an image by calling `OpenSimplexNoise.get_noise_2d()` for each coordinate, and generate a texture from that image. However, doing this in gdscript is very slow.

**Solution**
In this PR, I've updated `OpenSimplexNoise.get_image()` and `NoiseTexture` to accept a coordinate offset, so that they can generate noise images from anywhere in the noise space. In my tests, using the updated method is about an order of magnitude faster than using the gdscript method mentioned above. 

**Rationale**

- This enhancement enables the generation of large amounts of noise data.
- This enhancement is about 10x faster than doing the same thing in gdscript.
- This enhancement does not break compatibility.
- This enhancement is small and easily understood.

radishes committed:
Add support for generating noise images with an offset.

radishes commented:
**Demo**
Create a `Node2D` and attach the following script.

The default behavior is to generate a grid of sprites containing noise data using gdscript. This is the slow method which is compatible Godot 3.3.

Running `generate_noise_texture_grid()` or `generate_image_texture_grid()` demonstrate the new, faster functionality provided by this PR.

`extends Node2D

var grid_size = 4
var texture_size = 512
var noise = OpenSimplexNoise.new()

func _ready():
	generate_manually_offset_grid()
	#generate_noise_texture_grid()
	#generate_image_texture_grid()

func generate_manually_offset_grid():
	"""Generate noise textures by manually calculating offset in gdscript."""
	for grid_y in range(grid_size):
		for grid_x in range(grid_size):
			var pos = Vector2(grid_x, grid_y) * texture_size
			var data := PoolByteArray()
			data.resize(texture_size * texture_size)
			for y in range(texture_size):
				for x in range(texture_size):
					var n = noise.get_noise_2d(x + pos.x, y + pos.y) * 0.5 + 0.5
					data[y * texture_size + x] = n * 256
			var image := Image.new()
			image.create_from_data(texture_size, texture_size, false, Image.FORMAT_L8, data)
			var texture := ImageTexture.new()
			texture.create_from_image(image, 0)
			var sprite := Sprite.new()
			sprite.texture = texture
			sprite.position = pos
			add_child(sprite)

func generate_noise_texture_grid():
	"""Generate NoiseTextures using new offset feature."""
	for y in range(grid_size):
		for x in range(grid_size):
			var noise_texture := NoiseTexture.new()
			noise_texture.noise = noise
			noise_texture.width = texture_size
			noise_texture.height = texture_size
			var pos = Vector2(texture_size * x, texture_size * y)
			noise_texture.noise_offset = pos
			var sprite = Sprite.new()
			sprite.position = pos
			sprite.texture = noise_texture
			add_child(sprite)

func generate_image_texture_grid():
	"""Generate noise images using new offset feature and create texture manually."""
	for y in range(grid_size):
		for x in range(grid_size):
			var sprite = Sprite.new()
			var pos = Vector2(texture_size * x, texture_size * y)
			sprite.position = pos
			var image
			image = noise.get_image(texture_size, texture_size, pos.x, pos.y)
			var tex = ImageTexture.new()
			tex.create_from_image(image, 0)
			sprite.texture = tex
			add_child(sprite)`

radishes commented:
I will see about adding documentation for the new parameter today.

clayjohn commented:
This change would also have the side effect of fixing https://github.com/godotengine/godot/issues/47040

radishes commented:
This has been tagged "breaks compat" but I don't believe it does break compatibility. The default behavior remains the same as before.

radishes commented:
> 
> 
> This change would also have the side effect of fixing #47040

It wouldn't fix #47040 per se, as the (0, 0) noise space pixel would be generated as before, and because the default behavior is still to generate a noise image which spans (0, 0) - (width, height). But this enhancement would give users more flexibility in working around #47040.

Calinou commented on modules/opensimplex/doc_classes/NoiseTexture.xml:
```diff
@@ -31,6 +31,9 @@
 		<member name="noise" type="OpenSimplexNoise" setter="set_noise" getter="get_noise">
 			The [OpenSimplexNoise] instance used to generate the noise.
 		</member>
+    <member name="noise_offset" type="Vector2" setter="set_noise_offset" getter="get_noise_offset">
```
Use tab indentation:

```suggestion
    	<member name="noise_offset" type="Vector2" setter="set_noise_offset" getter="get_noise_offset">
```

radishes replied:
Fixed. Thanks for the feedback.

Calinou commented on modules/opensimplex/doc_classes/NoiseTexture.xml:
```diff
@@ -31,6 +31,9 @@
 			The [OpenSimplexNoise] instance used to generate the noise.
 		</member>
+    <member name="noise_offset" type="Vector2" setter="set_noise_offset" getter="get_noise_offset">
+			An offset used to specify the noise space coordinate of the top left corner of the generated noise. Defaults to (0, 0).
```
You don't need to add the default value in the description. Instead, `--doctool` should add the default value for you in an XML attribute. See [Contributing to the class reference](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html) in the documentation.

```suggestion
			An offset used to specify the noise space coordinate of the top left corner of the generated noise.
```

Calinou commented on modules/opensimplex/doc_classes/OpenSimplexNoise.xml:
```diff
@@ -30,9 +30,14 @@
 			<description>
-				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
+				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters. Optionally specify [code]noise_offset_x[/code] and [code]noise_offset_y[/code] values which are used as the coordinates of the
+        top left corner of the generated noise.
```
Don't wrap lines, as this creates new paragraphs in the editor help:

```suggestion
				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters. Optionally specify [code]noise_offset_x[/code] and [code]noise_offset_y[/code] values which are used as the coordinates of the top left corner of the generated noise.
```

radishes replied:
Fixed. Thanks for the feedback.

radishes commented on modules/opensimplex/noise_texture.cpp:
```diff
@@ -130,7 +134,7 @@ Ref<Image> NoiseTexture::_generate_texture() {
 		image = ref_noise->get_seamless_image(size.x);
 	} else {
-		image = ref_noise->get_image(size.x, size.y);
+		image = ref_noise->get_image(size.x, size.y, noise_offset.x, noise_offset.y);
```
This is the meat of the enhancement. Everything else is plumbing.

radishes commented on modules/opensimplex/open_simplex_noise.cpp:
```diff
@@ -96,15 +96,15 @@ void OpenSimplexNoise::set_lacunarity(float p_lacunarity) {
 	for (int i = 0; i < p_height; i++) {
 		for (int j = 0; j < p_width; j++) {
-			float v = get_noise_2d(j, i);
+			float v = get_noise_2d(j + p_offset_x, i + p_offset_y);
```
This is the meat of the enhancement. Everything else is plumbing.

Xrayez replied:
`get_noise_2d` functions accept `float` for x/y parameters, but `int p_offset_x, int p_offset_y` in `get_image()` accept integers. Would this be limited to integer offsets? Would it make any difference to use float offset over integer offset?

radishes replied:
I used `int` because the existing operation iterates over `int` values. What you are suggesting makes sense too, though. Do you think it would be valuable to have this accept `float` instead?

Xrayez replied:
Yeah, the loop operates on `int` values because noise is mapped to image coordinates, yet noise is not discrete compared to image, so it makes sense to me for this to accept `float` (rather, if we're talking about offset, perhaps it should use `real_t`, which is `double` in Godot by default, especially when we're talking about procedural generation with infinite worlds).

You could also change it to accept `Vector2`, there are many similar `offset` properties throughout Godot which are `Vector2`s, and because you're already using `Vector2` for `noise_offset` property in `NoiseTexture`, which would make the API consistent.

I'm not 100% sure about this, @JFonS may know better.

radishes replied:
Great idea. Thanks for the feedback. I've updated the PR to use `Vector2` as the type throughout.

Calinou commented on modules/opensimplex/doc_classes/OpenSimplexNoise.xml:
```diff
@@ -30,9 +30,14 @@
+      <argument index="2" name="noise_offset_x" type="int" default="0">
+      </argument>
+      <argument index="3" name="noise_offset_y" type="int" default="0">
+      </argument>
```
Use tab indentation:

```suggestion
			</argument>
			<argument index="2" name="noise_offset_x" type="int" default="0">
			</argument>
			<argument index="3" name="noise_offset_y" type="int" default="0">
			</argument>
```

radishes replied:
Fixed. Thanks for the feedback.

Calinou commented on modules/opensimplex/doc_classes/NoiseTexture.xml:
```diff
@@ -31,6 +31,9 @@
 			The [OpenSimplexNoise] instance used to generate the noise.
 		</member>
+		<member name="noise_offset" type="Vector2" setter="set_noise_offset" getter="get_noise_offset" default="Vector2( 0, 0 )">
+			An offset used to specify the noise space coordinate of the top left corner of the generated noise. This value is ignored if [code]seamless[/code] is enabled.
```
I didn't notice this at first, but you should use `[member]` references so they're clickable:

```suggestion
			An offset used to specify the noise space coordinate of the top left corner of the generated noise. [member noise_offset] is ignored if [member seamless] is [code]true[/code].
```

radishes replied:
Thanks for the heads up. I've updated this. I'll be sure to take a look at the documentation standards for future PRs.

Calinou commented on modules/opensimplex/doc_classes/OpenSimplexNoise.xml:
```diff
@@ -31,8 +31,12 @@
+			</argument>
 			<description>
-				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
+				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters. Optionally specify [code]noise_offset_x[/code] and [code]noise_offset_y[/code] values which are used as the coordinates of the top left corner of the generated noise.
```
```suggestion
				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters. If [code]noise_offset_x[/code] and [code]noise_offset_y[/code] are specified, these offset values are used as the coordinates of the top-left corner of the generated noise.
```

Calinou commented on modules/opensimplex/doc_classes/OpenSimplexNoise.xml:
```diff
@@ -31,8 +31,12 @@
+			</argument>
 			<description>
-				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
+				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [member width] and [member height], based on the current noise parameters. If [member noise_offset_x] and [member noise_offset_y] values are specified, these offset values are used as the coordinates of the top-left corner of the generated noise.
```
Method arguments aren't specifiable as `[member]`, so you need to use `[code]`:

```suggestion
				Generate a noise image in [constant Image.FORMAT_L8] format with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters. If [code]noise_offset_x[/code] and [code]noise_offset_y[/code] are specified, these offset values are used as the coordinates of the top-left corner of the generated noise.
```

radishes replied:
Fixed, and I'm looking into the pre-commit hooks now. Thank you for the feedback and sorry for the churn.

radishes commented:
> 
> 
> > This change would also have the side effect of fixing #47040
> 
> It wouldn't fix #47040 per se, as the (0, 0) noise space pixel would be generated as before, and because the default behavior is still to generate a noise image which spans (0, 0) - (width, height). But this enhancement would give users more flexibility in working around #47040.

At the suggestion of @Xrayez I updated this PR to use `Vector2` all the way through. This gives an effective way to work around #47040: pass non-integer offsets, for example (0.5, 0.5), and the noise map will not generate a value for (0, 0).

Xrayez commented on modules/opensimplex/open_simplex_noise.h:
```diff
@@ -75,7 +75,7 @@ class OpenSimplexNoise : public Resource {
 	float get_lacunarity() const { return lacunarity; }
 
-	Ref<Image> get_image(int p_width, int p_height) const;
+	Ref<Image> get_image(int p_width, int p_height, Vector2 p_noise_offset) const;
```
Nitpick but prefer to pass by reference rather than by value (see the same usage as in `get_noise_*` functions):

```suggestion
	Ref<Image> get_image(int p_width, int p_height, const Vector2 &p_noise_offset) const;
```

radishes replied:
Good feedback, thanks. I've checked in an update.

clayjohn commented:
Looks great! Can you please [squash](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) your commits into one. We like to keep our git history very clean. After that, it should be ready to merge!

radishes commented:
> 
> 
> Looks great! Can you please [squash](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) your commits into one. We like to keep our git history very clean. After that, it should be ready to merge!

Done. Thank you for the link!

akien-mga commented:
FYI, your local Git author details (see header of https://github.com/godotengine/godot/commit/f33800a9c8023843de271280a7482a5b7911e8e6.patch) don't seem to match the email registered in your GitHub account, which is why the commit doesn't appear as authored by "@radishes", so your GitHub account would not be properly credited as contributor once merged. It's not a problem technically for Git or GitHub, but if you want to claim this commit to your account, you can add this email as secondary email in your GitHub settings.

radishes commented:
> 
> 
> FYI, your local Git author details (see header of https://github.com/godotengine/godot/commit/f33800a9c8023843de271280a7482a5b7911e8e6.patch) don't seem to match the email registered in your GitHub account, which is why the commit doesn't appear as authored by "@radishes", so your GitHub account would not be properly credited as contributor once merged. It's not a problem technically for Git or GitHub, but if you want to claim this commit to your account, you can add this email as secondary email in your GitHub settings.

Thank you! I've added that email address as a secondary email to my Github account.

JFonS reviewed (approved):
Looks good to me :)

akien-mga commented on modules/opensimplex/noise_texture.h:
```diff
@@ -56,6 +56,7 @@ class NoiseTexture : public Texture2D {
 
 	Ref<OpenSimplexNoise> noise;
 	Vector2i size = Vector2i(512, 512);
+	Vector2 noise_offset = Vector2(0, 0);
```
```suggestion
	Vector2 noise_offset;
```
`Vector2` is constructed with `0, 0` values out of the box.

akien-mga commented on modules/opensimplex/open_simplex_noise.cpp:
```diff
@@ -161,7 +161,7 @@ void OpenSimplexNoise::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_lacunarity"), &OpenSimplexNoise::get_lacunarity);
 
-	ClassDB::bind_method(D_METHOD("get_image", "width", "height"), &OpenSimplexNoise::get_image);
+	ClassDB::bind_method(D_METHOD("get_image", "width", "height", "noise_offset"), &OpenSimplexNoise::get_image, DEFVAL(Vector2(0, 0)));
```
Can be simplified to:
```suggestion
	ClassDB::bind_method(D_METHOD("get_image", "width", "height", "noise_offset"), &OpenSimplexNoise::get_image, DEFVAL(Vector2()));
```

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

akien-mga commented:
This was label as a candidate to cherry-pick for `3.x`, but there's an issue due to #30424 which was not cherry-picked as it breaks compatibility. So I'm not sure if adding this feature as is would make sense or if the x and y coordinates should also be swapped, but that potentially makes things even more confusing.

radishes commented:
I implemented the noise offset changes in the 3.x branch for myself and noticed the swapped axes bug too. I left the bug alone and added the offsets as shown below.

The method I pasted below produces a good outcome because, despite the coordinate bug, it produces noise images which can be tiled together for a seamless view of the larger noise space. This satisfies my use cases, and probably most use cases; I'm not concerned about the unexpected orientation of the generated noise.

However, if a user is generating noise images this way, and comparing specific areas with the equivalent noise generated using `OpenSimplexNoise.get_noise_2d()`, they will notice the discrepancy between the two. But, since `get_image()` already causes this discrepancy, I don't see any harm in leaving the bug alone to maintain compatibility.

tl;dr: I recommend implementing it in 3.x as below, for the "principal of least surprise" reasons. This is compatible with expected outcomes considering #30424. If #30424 gets fixed, it will still work with the new noise offset feature. The noise images produced are continuous and tile-able in either case.

```
Ref<Image> OpenSimplexNoise::get_image(int p_width, int p_height, const Vector2 &p_noise_offset) const {
	PoolVector<uint8_t> data;
	data.resize(p_width * p_height);

	PoolVector<uint8_t>::Write wd8 = data.write();

	for (int i = 0; i < p_height; i++) {
		for (int j = 0; j < p_width; j++) {
			float v = get_noise_2d(float(j) + p_noise_offset.x, float(i) + p_noise_offset.y);
			v = v * 0.5 + 0.5; // Normalize [0..1]
			wd8[(i * p_width + j)] = uint8_t(CLAMP(v * 255.0, 0, 255));
		}
	}

	Ref<Image> image = memnew(Image(p_width, p_height, false, Image::FORMAT_L8, data));
	return image;
}
```

radishes commented:
I think swapping the offset coordinates to match the bug is the least surprising outcome for users. The example below produces tile-able continuous noise. Maybe an explanatory comment would help?

```float v = get_noise_2d(float(i) + p_noise_offset.y, float(j) + p_noise_offset.x);  // Axes swapped to maintain compatibility, per #30424. ```

Xrayez commented:
In any case, #30424 is a clear bug to me which could've been cherry-picked as well, but I guess too many people might already rely on the wrongly swapped coordinates...

radishes commented:
If there's going to be a 3.4, maybe that's a good time to break compatibility and fix the bug? Is that option on the table?

Xrayez commented:
> If there's going to be a 3.4, maybe that's a good time to break compatibility and fix the bug? Is that option on the table?

I personally think it would be worth it, as long as the change is documented of course. But it's up to Akien to decide.

akien-mga commented:
I agree, I cherry-picked #30424 for 3.4.

> We've had regressions in `NoiseTexture` due to _major_ thread modernization backports in the past: #46907 (and many other regressions because of #45618), so I don't see a reason why such a _minor_ bug fix cannot be cherry-picked...

That's not a very good argument. Those regressions have been fixed before the stable release, so there's no compat breakage in that specific instance. On the other hand, the now cherry-picked #30424 *will* break compat in a stable release. That's apples and oranges.

akien-mga commented:
Cherry-picked for 3.4.
